include/up-cpp/datamodel/builder/Uuid.h

```diff
@@ -0,0 +1,154 @@
+///     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///     |var|                          rand_b                           |
+///     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///     |                           rand_b                              |
+///     +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///
+/// | Field      | RFC2119 |
+/// | -----      | --------|
+/// | unix_ts_ms | 48 bit big-endian unsigned number of Unix epoch timestamp in
+///                milliseconds as per Section 6.1  of RFC |
+/// | ver        | MUST be 8 per Section 4.2 of draft-ietf-uuidrev-rfc4122bis |
+/// | counter    | MUST be a 12 bit counter field that is reset at each
+///                unix_ts_ms tick, and incremented for each UUID generated
+///                within the 1ms precision of unix_ts_ms The counter provides
+///                the ability to generate 4096 events within 1ms however the
+///                precision of the clock is still 1ms accuracy |
+/// | var        | MUST be the The 2 bit variant defined by Section 4.1 of RFC |
+/// | rand_b     | MUST 62 bits random number that is generated at
+/// initialization
+///                time of the uE only and reused otherwise |
+struct UuidBuilder {
+	/// @brief Get a UuidBuilder in the default, production mode.
+	///
+	/// @remarks This should be used in most cases.
+	///
+	/// @returns A UuidBuilder in the default, production mode.
+	static UuidBuilder getBuilder();
+
+	/// @brief Get a UuidBuilder in the test mode.
+	///
+	/// The testing mode of UuidBuilder allows for the time and random sources
+	/// to be replaced such that deterministic tests can be written in
+	/// situations where the normal behavior of UuidBuilder would interfere.
+	///
+	/// @remarks The provided UuidBuilder starts with an identical state to
+	///          one returned by getBuilder() with one difference: it will
+	///          allow customization of its behavior through the `withX()`
+	///          interfaces.
+	///
+	/// @returns A UuidBuilder in the test mode.
+	static UuidBuilder getTestBuilder();
+
+	/// @brief Sets the time source for a UuidBuilder in test mode.
+	///
+	/// @post All built UUIDs will use the provided function to get time
+	///       values instead of calling `std::chrono::system_clock::now()`
+	///
+	/// @note This can only be used with a UuidBuilder created with the
+	///       getTestBuilder() interface.
+	///
+	/// @param A callable that returns a time_point. Will be called whenever
+	///        build() is called to populate the time field in the built UUID.
+	///
+	/// @throws std::domain_error If called on a non-test UuidBuilder.
+	/// @returns A reference to thus UuidBuilder.
+	UuidBuilder& withTimeSource(
+	    std::function<std::chrono::system_clock::time_point()>&&);
+
+	/// @brief Sets the random value source for a UuidBuilder in test mode.
```

Comment:
This is for testing and special uses, similar to `withTimeSource()`. Any attempt to use this will throw an exception unless `getTestBuilder()` is called. It allows for deterministic behavior when required in a test environment. We could further protect this by not defining `getTestBuilder()` in the main library code such that casual users of the library won't be able to link to it.